docs/context.rst

```diff
@@ -300,28 +302,200 @@ DetDb auto-documentation
+elevation of the scan, whether the observation was done during the day
+or night, etc.).  The ObsDb is also used, by the Context system, to
+select the appropriate metadata results from a metadata archive (for
+example, the timestamp associated with an observation could be used to
+find an appropriate pointing offset).
+
+The ObsDb is constructed from two tables.  The "obs table" contains
+one row per observation and is appropriate for storing basic
+descriptive data about the observation.  The "tags table" associates
+observations to particular labels (called tags) in a many-to-many
+relationship.
+
+The ObsDb code does not really require that any particular information
+be present; it does not insist that there is a "timestamp" field, for
+example.  Instead, the ObsDb can contain information that is needed in
+a specific context.  However, recommended field names for some typical
+information types are given in `Standardized ObsDb field names`_.
+
+.. note::
+
+   The difference between ObsDb and ObsFileDb is that ObsDb contains
+   arbitrary high-level descriptions of Observations, while ObsFileDb
+   only contains information about how the Observation data is
+   organized into files on disk.  ObsFileDb is consulted when it is
+   time to load data for an observation into memory.  The ObsDb is a
+   repository of information about Observations, independent of where
+   on disk the data lives or how the data files are organized.
+
+Creating an ObsDb
+=================
+
+To create an ObsDb, one must define columns for the obs table, and
+then add data for one or more observations, then write the results to
+a file.
+
+Here is a short program that creates an ObsDb with entries for two
+observations::
+
+  from sotodlib.core import metadata
+
+  # Create a new Db and add two columns.
+  obsdb = metadata.ObsDb()
+  obsdb.add_obs_columns(['timestamp float', 'hwp_speed float', 'drift string'])
+
+  # Add two rows.
+  obsdb.update_obs('myobs0', {'timestamp': 1900000000.,
+                              'hwp_speed': 2.0,
+                              'drift': 'rising'})
+  obsdb.update_obs('myobs1', {'timestamp': 1900010000.,
+                              'hwp_speed': 1.5,
+                              'drift': 'setting'})
+
+  # Apply some tags (this could have been done in the calls above
+  obsdb.update_obs('myobs0', tags=['hwp_fast', 'cryo_problem'])
+  obsdb.update_obs('myobs1', tags=['hwp_slow'])
+
+  # Save (in gzip format).
+  obsdb.to_file('obsdb.gz')
+
+
+The column definitions must be specified in a format compatible with
+sqlite; see :py:meth:`ObsDb.add_obs_columns`.  When the user adds data
+using :py:meth:`ObsDb.update_obs`, the first argument is the
+``obs_id``.  This is the primary key in the ObsDb and is also used to
+identify observations in the ObsFileDb.  When we write the database
+using :py:meth:`ObsDb.to_file`, using a .gz extension selects gzip
+output by default.
+
+
+Using an ObsDb
```

Comment:
Your ObsDB and ObsFileDB are essentially "Object Relational Managers" (ORMs).  In Python, there are several popular tools that represent a class as a database table and each row is an instance of that class.  For example, with Django you define a "Model" (https://docs.djangoproject.com/en/3.0/topics/db/models/) which is a class representing a table and then add rows (instances) and make queries (https://docs.djangoproject.com/en/3.0/topics/db/queries/) while also applying filters on the values.  And you can do joins between models as well.  Most of these tools support multiple backends (sqlite, postgresql, etc).

Of course you can "roll your own" as you have done here, but just wondering if there was a particular reason you rejected an off-the-shelf tool for this?

Reply:
For ObsFileDb you are probably right... I have not looked at that one closely, lately, but the table structure is ~completely specified in code.  But for ObsDb I'm less sure.

The ObsDb schema is deliberately under-specified.  The code that reads an ObsDb will not know, in advance, what the columns of the obs table are (other than obs_id).  Perhaps the ORMs can handle that (not knowing column names / instance variables in advance) gracefully, but it does not seem to be their focus.  I also like that it so easy to work with the table of results one gets back; the ResultSet exposes the data as numpy arrays and dicts, so they are ready to go for whatever.

Reply:
Sounds good.  And these are relatively straightforward classes representing small databases.  If you needed to also target things like postgres due to the database size, that might be a stronger argument for using off-the-shelf packages.